torch/fx/subgraph_rewriter.py

```diff
@@ -0,0 +1,202 @@
+from torch.fx import Graph, Node, symbolic_trace
+
+import collections
+from typing import Callable, Dict, List, Set
+
+# `anchor` - node from which the match was found
+# `nodes_map` - maps nodes in the pattern subgraph to nodes in the
+#               larger graph
+Match = collections.namedtuple("Match", "anchor nodes_map")
+
+class SubgraphMatcher:
+    def __init__(self, pattern: Graph) -> None:
+        self.pattern = pattern
+        try:
+            self.pattern_anchor = next(iter(reversed(pattern.nodes)))
+        except Exception:
+            raise ValueError("SubgraphMatcher cannot be initialized with an "
+                             "empty pattern")
+        # Maps nodes in the pattern subgraph to nodes in the larger graph
+        self.nodes_map: Dict[Node, Node] = {}
+
+    def matches_subgraph_from_anchor(self, anchor: Node) -> bool:
+        """
+        Check if the whole pattern can be matched starting from `anchor` in
+        the larger graph
+        """
+        self.nodes_map = {}
+        if (self.match_nodes(self.pattern_anchor, anchor)):
+            return True
+        return False
+
+    def match_nodes(self, n1: Node, n2: Node) -> bool:
+        """
+        Compare Nodes `n1` (pattern node) and `n2` (graph node) for equality
+        """
```

Comment:
Is this meant to be called externally or is it just factored for the recursion? If the latter, lets make this private (`_match_nodes`) and move the docstring text here up to the docstring for `matches_subgraph_from_anchor`